test/middleware/routing_middleware_test.dart

```diff
@@ -0,0 +1,423 @@
+        });
+
+        final pipeline = middleware1.meddle(middleware2.meddle(
+            (final RequestContext ctx) async =>
+                (ctx as RespondableContext).withResponse(FakeResponse(404))));
+
+        final initialCtx = FakeRequest('/router1/apple').toContext(Object());
+        final resultingCtx = await pipeline(initialCtx);
+        final response =
+            (resultingCtx as ResponseContext).response as FakeResponse;
+
+        expect(handler1Called, isTrue);
+        expect(handler2Called, isFalse);
+        expect(response.statusCode, equals(201));
+        expect(params1, equals({#item: 'apple'}));
+        expect(response.parameters, equals({#item: 'apple'}));
+      });
+
+      test(
+          'Given two RoutingMiddleware instances in a pipeline, '
+          'When a request matches a route in the second router (but not the first), '
+          'Then the handler from the second router is executed with correct parameters',
+          () async {
+        Map<Symbol, String>? params2;
+        bool handler1Called = false;
+        bool handler2Called = false;
+
+        router1.add(Method.get, '/router1/:item',
+            (final RequestContext ctx) async {
+          handler1Called = true;
+          return (ctx as RespondableContext).withResponse(FakeResponse(201));
+        });
+        router2.add(Method.get, '/router2/:data',
+            (final RequestContext ctx) async {
+          handler2Called = true;
+          params2 = ctx.pathParameters;
+          return (ctx as RespondableContext)
+              .withResponse(FakeResponse(202, parameters: params2));
+        });
+
+        final pipeline = middleware1.meddle(middleware2.meddle(
+            (final RequestContext ctx) async =>
+                (ctx as RespondableContext).withResponse(FakeResponse(404))));
+
+        final initialCtx = FakeRequest('/router2/banana').toContext(Object());
+        final resultingCtx = await pipeline(initialCtx);
+        final response =
+            (resultingCtx as ResponseContext).response as FakeResponse;
+
+        expect(handler1Called, isFalse);
+        expect(handler2Called, isTrue);
+        expect(response.statusCode, equals(202));
+        expect(params2, equals({#data: 'banana'}));
+        expect(response.parameters, equals({#data: 'banana'}));
+      });
+
+      test(
+          'Given two RoutingMiddleware instances in a pipeline, '
+          'When a request does not match any route in either router, '
+          'Then the final next handler is called', () async {
+        bool handler1Called = false;
+        bool handler2Called = false;
+        bool fallbackCalled = false;
+
+        router1.add(Method.get, '/router1/:item',
+            (final RequestContext ctx) async {
+          handler1Called = true;
+          return (ctx as RespondableContext).withResponse(FakeResponse(201));
+        });
+        router2.add(Method.get, '/router2/:data',
```

Comment:
Is there any specificity matching (in and across routers), like `/x/:foo` vs. `/x/foo` (variable vs. literal?)

Maybe not matter what's currently implemented, do we have a test for the behavior, so that we would notice changes?

Reply:
Yes, literal matching takes precedence. It is tested yes